### ISSUE TITLE
Wasm-parser gives "invalid UTF-8 encoding" on intermediate output from wasm-bindgen

### ISSUE THREAD

**Issue by TimHambourger** [xtuc/webassemblyjs#414]
I'm filing this to track the remaining [failing cases](https://github.com/xtuc/webassemblyjs/issues/407#issuecomment-404995901) from #407 that aren't addressed by #413. From now on I'll consider these separate from #407.

Again, those cases are
```
get37_debug.wasm -- addAssignExports -- FAIL
    wasm-parser error:        Error: invalid UTF-8 encoding
    native WebAssembly error: null
get37_debug.wasm -- oneCharExports -- FAIL
    wasm-parser error:        Error: invalid UTF-8 encoding
    native WebAssembly error: null
get37_debug.wasm -- removeExports -- FAIL
    wasm-parser error:        Error: invalid UTF-8 encoding
    native WebAssembly error: null
get37_debug.wasm -- renameImports -- FAIL
    wasm-parser error:        Error: invalid UTF-8 encoding
    native WebAssembly error: null
get37_no_struct_debug.wasm -- addAssignExports -- FAIL
    wasm-parser error:        Error: invalid UTF-8 encoding
    native WebAssembly error: null
get37_no_struct_debug.wasm -- oneCharExports -- FAIL
    wasm-parser error:        Error: invalid UTF-8 encoding
    native WebAssembly error: null
get37_no_struct_debug.wasm -- removeExports -- FAIL
    wasm-parser error:        Error: invalid UTF-8 encoding
    native WebAssembly error: null
get37_no_struct_debug.wasm -- renameImports -- FAIL
    wasm-parser error:        Error: invalid UTF-8 encoding
    native WebAssembly error: null
get37_no_struct_release.wasm -- addAssignExports -- FAIL
    wasm-parser error:        Error: invalid UTF-8 encoding
    native WebAssembly error: null
get37_no_struct_release.wasm -- oneCharExports -- FAIL
    wasm-parser error:        Error: invalid UTF-8 encoding
    native WebAssembly error: null
get37_no_struct_release.wasm -- removeExports -- FAIL
    wasm-parser error:        Error: invalid UTF-8 encoding
    native WebAssembly error: null
get37_no_struct_release.wasm -- renameImports -- FAIL
    wasm-parser error:        Error: invalid UTF-8 encoding
    native WebAssembly error: null
get37_release.wasm -- addAssignExports -- FAIL
    wasm-parser error:        Error: invalid UTF-8 encoding
    native WebAssembly error: null
get37_release.wasm -- oneCharExports -- FAIL
    wasm-parser error:        Error: invalid UTF-8 encoding
    native WebAssembly error: null
get37_release.wasm -- removeExports -- FAIL
    wasm-parser error:        Error: invalid UTF-8 encoding
    native WebAssembly error: null
get37_release.wasm -- renameImports -- FAIL
    wasm-parser error:        Error: invalid UTF-8 encoding
    native WebAssembly error: null
```

One thing to note: all of these files are intermediate outputs from [wasm-bindgen](https://github.com/rustwasm/wasm-bindgen). They represent the step after Rust compilation but before processing by the wasm-bindgen CLI. As I understand it, wasm-bindgen stashes extensive metadata in that intermediate output (I think in the data section) that then gets interpreted and stripped by the CLI. So my first guess is that that intermediate metadata is involved.

**Comment by xtuc** [comment]
> As I understand it, wasm-bindgen stashes extensive metadata in that intermediate output (I think in the data section) 

Yes, that's what I had in mind too. I think it's in new custom sections instead of data, which we should ignore (apart from `name` custom section).

**Comment by TimHambourger** [comment]
@xtuc Ah, thanks. A couple more observations:

1. I think wasm-edit is a red herring. My test suite uses different decode options for the [initial decode](https://github.com/TimHambourger/webassemblyjs-wasm-corruption/blob/c7c4ef0fb605ca5852a9dd1ba788d5f8fb3f43f3/webassemblyjs-repro/index.js#L131-L137) vs the [2nd decode](https://github.com/TimHambourger/webassemblyjs-wasm-corruption/blob/c7c4ef0fb605ca5852a9dd1ba788d5f8fb3f43f3/webassemblyjs-repro/index.js#L71) that verifies wasm-edit output. If I attempt the initial `decode` with the default options, it throws the same error. I'll update the name of this issue to reflect this.
1. The error's getting thrown while wasm-parser is parsing the custom name section. I don't know whether that means the other custom sections are irrelevant or if there's something more complicated going on.
1. The `decode` option that avoids this is `ignoreDataSection: true`. I observed that `ignoreDataSection` bypasses parsing of the custom name section as well as the data section. Apparently,`ignoreCustomNameSection: true` does not avoid parsing of the custom name section but instead gets picked up downstream of the parse.

I'll keep debugging and let you know what I find.

**Comment by xtuc** [comment]
> I observed that ignoreDataSection bypasses parsing of the custom name section [...]

That doesn't seem right to me, that's a bug.

> Apparently, `ignoreCustomNameSection: true` does not avoid parsing of the custom name section [...]

Here's the code: https://github.com/xtuc/webassemblyjs/blob/master/packages/wasm-parser/src/index.js#L144 which only disable name resolution. The naming is very confusing.

**Comment by TimHambourger** [comment]
@xtuc. Ah ok, I think I have this figured out.

First on `ignoreDataSection`: Looks like this is just [a typo](https://github.com/xtuc/webassemblyjs/blob/56bbd621a92690bd86aa4093bf13ec5d5cb799f4/packages/wasm-parser/src/decoder.js#L1712). I think that should be `numberOfElementsu32.nextIndex` instead of `numberOfElements.nextIndex`. As is, `numberOfElements` is a number not an object, so `numberOfElements.nextIndex` evaluates to undefined, `remainingBytes` gets initialized to `NaN`, and the `eatBytes` call ends up setting the current `offset` to `NaN`. That causes [the test](https://github.com/xtuc/webassemblyjs/blob/56bbd621a92690bd86aa4093bf13ec5d5cb799f4/packages/wasm-parser/src/decoder.js#L1777) for whether to keep parsing to evaluate to false, effectively skipping the rest of the input. It's only noticeable if you have custom sections after the data section.

As far as the invalid UTF-8 error: This appears to be due to a [byte arithmetic error](https://github.com/xtuc/webassemblyjs/blob/56bbd621a92690bd86aa4093bf13ec5d5cb799f4/packages/wasm-parser/src/decoder.js#L1738). `sectionSizeInBytes - sectionName.nextIndex` is not a valid way to compute `remainingBytes` b/c `readUTF8String` includes an [eatBytes call](https://github.com/xtuc/webassemblyjs/blob/56bbd621a92690bd86aa4093bf13ec5d5cb799f4/packages/wasm-parser/src/decoder.js#L208). So just using the `nextIndex` from `readUTF8String` doesn't include the bytes that were eaten by `readUTF8String` itself. In the case of the custom name section, the LEB-encoded string length takes 1 byte (length of "name" is `0x04`), so `remainingBytes` ends up being 1 byte too large. The result is that [parseNameSection](https://github.com/xtuc/webassemblyjs/blob/56bbd621a92690bd86aa4093bf13ec5d5cb799f4/packages/wasm-parser/src/decoder.js#L1164-L1179) ends up trying to parse one more subsection than it ought too.

If the custom name section is the last section in the input, this ends up being inert. The first step to parsing a subsection is to [read a byte](https://github.com/xtuc/webassemblyjs/blob/56bbd621a92690bd86aa4093bf13ec5d5cb799f4/packages/wasm-parser/src/decoder.js#L1169) to determine which subsection it is. If wasm-parser has already reached the end of the input buffer, then `sectionTypeByte` will evaluate to `undefined`. Since that fails all of the explicit comparisons below, the extra pass through the while loop is a no-op.

But in the case of the intermediate wasm-bindgen outputs, the custom name section is NOT the last section in the input since it's followed by wasm-bindgen's custom __wasm_bindgen_unstable section. B/c of the extra byte, wasm-parser continues trying to parse the next custom section as if it were another name subsection, and that's when the UTF-8 error gets thrown. If you break on the failing `utf8.decode` call and inspect `bytes`, it includes the fragment "__wasm_bindgen_unstable" prepended w/ a few extra bytes and followed by a lot of JSON.

So, bottom line, these should both be easy fixes. I'll submit PRs.

**Comment by xtuc** [comment]
Good catch @TimHambourger! I already submited a PR https://github.com/xtuc/webassemblyjs/pull/416 for the `remainingBytes` issue.

**Comment by TimHambourger** [comment]
@xtuc Awesome! Just submitted #427 which corrects the `remainingBytes` for custom sections. Thanks for the quick turnaround on #416!

**Comment by xtuc** [comment]
Thanks! Also here's a fix on the custom name section: https://github.com/xtuc/webassemblyjs/pull/418. That should remove a couple of bugs too.

**Comment by xtuc** [comment]
This has been fixed on master.

**Comment by TimHambourger** [comment]
@xtuc. Can you please re-open? It looks the exception is gone in current master, but only for incidental reasons. The underlying bug (that wasm-parser reads past the end of the custom name section into the next section) is still present. That fix needs my #427 or something like it.

The reason the exception is gone is b/c parsing of the module name subsection is [temporarily disabled](https://github.com/xtuc/webassemblyjs/blob/5247381432319ecd9d64351060299c259f18b543/packages/wasm-parser/src/decoder.js#L1191-L1214). Due to the bug tracked here, when wasm-parser encounters another custom section after the custom name section, it mistakes it for the module name subsection. That's b/c custom sections and the module name subsection both start with `0x00`. With module name parsing disabled, that means the following custom section gets skipped (wasm-parser follows the default branch in the linked snippet). When module name parsing is re-enabled, wasm-parser will instead try to parse the following section as the module name subsection. I'd expect that to throw an exception, which will get caught [here](https://github.com/xtuc/webassemblyjs/blob/5247381432319ecd9d64351060299c259f18b543/packages/wasm-parser/src/decoder.js#L1782-L1790) and cause the metadata from the custom name section to be ignored.

**Comment by xtuc** [comment]
Ok, you are right. Sorry I was probably too much in a hurry.
